### ISSUE TITLE
Design of the language structures and the visitor

### ISSUE THREAD

**Issue by flip1995** [rust-linting/design#12]
I thought a little bit about how to implement the language structures. Meaning expressions, statements, items, ...

I would only create one level of those structures and not split it up into AST/HIR/... levels like it is done in the compiler. (I think we already agreed on this somewhere?). That also implies that we don't implement something like `Early`- and `LateLintPass`es, but just one visitor pattern.

So what should we base this implementation on? I think a good reference for the implementation would be if we follow the [Rust Reference](https://doc.rust-lang.org/stable/reference/) and how it defines expressions, statements, etc.

On the plus side of this approach, I see that the reference is the most compiler-neutral thing that currently exists, where we could base this on. 

One concern that came to mind was, that it describes the language on a very high level. So if we would follow along tightly, we'd also for example differentiate between `while` loops and `for` loops, which isn't done in the HIR. I'm not sure if this is good or bad for the end user of the API.

**Comment by jhpratt** [comment]
HIR doesn't differentiate between `while` and `for` loops? I thought that distinction was removed when lowered to MIR. I _would_ like to be able to differentiate between the two.

I'll be honest, I don't quite understand the separation between early and late passes to begin with, so I'm probably not the best person to judge whether it's a good idea to go along that line.

**Comment by DevinR528** [comment]
HIR can distinguish between `for` and `while` if we are talking about [`Expr::Loop {...}`](https://doc.rust-lang.org/stable/nightly-rustc/rustc_hir/hir/enum.ExprKind.html#variant.Loop) (`LoopSource` enum).

The difference is early is AST so you get parsed source code. All macro calls are unexpanded all sugar is unexpanded also i.e. `?` or `if let ...` and loops. Late has everything expanded but most (if not all of the AST) info is kept. You can find the macro that a section of code was expanded from or tell if a `match` is expanded from an `if let`. You also get all the type info and context in late, early is like `syn` there is much less context and no type info.

**Comment by jhpratt** [comment]
Ah, gotcha. It's probably best to not explicitly provide that distinction, but I _would_ like to have the ability to do what is functionally possible in either. Namely syntax-only lints should be possible, while I believe it would be useful to lint both pre and post-macro expansion.

**Comment by flip1995** [comment]
> The difference is early is an AST so you get parsed source code. All macro calls are unexpanded

That's not 100% true. In the AST you get with `EarlyLintPass` all macro calls are expanded, with information attached to the spans if they come from a macro call or other expansions. See [`ExpnKind`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_expand/hygiene/enum.ExpnKind.html) for the possible expansions at different stages in the compile process.

> while I believe it would be useful to lint both pre and post-macro expansion.

pre-macro expansion is difficult, since rustc currently doesn't provide a hook for getting this information (there is the deprecated pre-expansion pass, but we shouldn't rely on it to exist forever).

Maybe this will be easier with macros 2.0, but that's a long way down the road...

**Comment by jhpratt** [comment]
Does rustc provide a way to know if code was macro-generated? It would be best to not trigger lints on code generated by external macros for obvious reasons.

**Comment by flip1995** [comment]
> Does rustc provide a way to know if code was macro-generated?

Yes, every expansion is tracked in rustc, so you can always know if code was produced by expansion (and which expansion) or by the programmer.

**Comment by flip1995** [comment]
> I think a good reference for the implementation would be if we follow the [Rust Reference](https://doc.rust-lang.org/stable/reference/) and how it defines expressions, statements, etc.

What do you think about basing the API on the Rust reference?

**Comment by jplatte** [comment]
> Yes, every expansion is tracked in rustc, so you can always know if code was produced by expansion (and which expansion) or by the programmer.

Is it really that simple? I remember Clippy often linting on macro-generated code and the fix always seemed to be checking the span of the generated code, which would only fix it for code with the macro invocation as the span, i.e. most `quote!`d code but *not* code using `quote_spanned!` + a span from the macro input.

**Comment by flip1995** [comment]
> Is it really that simple? I remember Clippy often linting on macro-generated code and the fix always seemed to be checking the span of the generated code

Yes, but that is quite simple to do by checking the `ExpnData` of a span.

> _not_ code using `quote_spanned!` + a span from the macro input.

Doing this basically manually changes the `ExpnData`. You're explicitly telling the compiler that the produced code comes from a different location. There's nothing you can do about that in static analysis after the fact.

**Comment by jhpratt** [comment]
> What do you think about basing the API on the Rust reference?

Seems reasonable upon first thought. Picking from the reference and syn could plausibly lead to something useful.

(I'm not familiar with the intricacies of expansion, so I'm staying aside here)

**Comment by DevinR528** [comment]
I think going from the reference + spans + some Id type would work. I did start this https://github.com/rust-linting/api-types if looking at code helps anyone else.
